File: solutions/ocaml/047-rotate-list.ml
(*
  Problem 47: Rotate List (LeetCode 61)
  Difficulty: Med
  Language: OCaml
*)
let rotate_right lst k =
  match lst with
  | [] -> []
  | _ ->
    let n = List.length lst in
    let rot = k mod n in
    if rot = 0 then lst
    else
      let rec split i acc = function
        | [] -> (List.rev acc, [])
        | x :: xs -> if i = 0 then (List.rev acc, x :: xs)
                     else split (i-1) (x :: acc) xs
      in
      let (front, back) = split (n - rot) [] lst in
      back @ front

File: solutions/rust/047-rotate-list.rs
/**
 * Problem 47: Rotate List (LeetCode 61)
 * Difficulty: Med
 * Language: Rust
 */
pub fn rotate_right(list: Vec<i32>, k: i32) -> Vec<i32> {
    if list.is_empty() { return vec![]; }
    let n = list.len();
    let rot = (k as usize) % n;
    if rot == 0 { return list; }
    [&list[n - rot..], &list[..n - rot]].concat()
}
